### ISSUE TITLE
[BUG]Engine image stuck in deploying `Text file busy`

### ISSUE THREAD

**Issue by dstuart** [longhorn/longhorn#1116]
**Describe the bug**
Longhorn engine stuck in deploying. We have a 5 node cluster and recently upgrade to v0.8.0 all seemed fine but the longhorn engine is stuck in deploying. 

One of the engine-images is successfully running but the rest are crash back with the log message cp: cannot create regular file '/data/longhorn': Text file busy
Looking the filesystem it seems like the ones that have failed arent creating an engine in /var/lib/rancher/longhorn/engine-binaries but im not sure if thats the reason. I have tried removing the container image and pulling again but not success. Any ideas? 

We are using rancher v2.3.5

**Comment by yasker** [comment]
@dstuart How long it has been like that? Normally the issue should be self-solved after a while. Also, which engine image caused the trouble? If it's v0.7.0, you can scale down the workload and it should detach the volume, then you can upgrade the engine and scale up the workload again to complete the upgrade process.

**Comment by dstuart** [comment]
I left it for a 3 hours and it never righted had about 35 restarts. I am using v0.8.0 I tried a redeploy but couldn't scale up and down. I also change it to deploy on a single node which seemed to work in that it said the engine was running but it still said deploying in longhorn UI and I couldnt create a PVC

**Comment by dstuart** [comment]
I managed to get it running on one of the nodes but as it requires 100% availability it doesn't need to work. On the nodes that didn't have the engine directory created on container deploy i.e. var/lib/rancher/longhorn/engine-binaries/longhornio-longhorn-engine-v0.8.0 I tried to tar it up from my one working node and move it onto all the other nodes but it didn't work and still complained about  "p: cannot create regular file '/data/longhorn': Text file busy".

**Comment by yasker** [comment]
@dstuart `/data/longhorn` is the Longhorn engine binary that needs to be installed to `/var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v0.8.0`, so it can be used by the instance managers to start engine/replica processes. I think there are some running workloads is using the binary, though if that's the case, I don't know why this binary image has been deployed twice.

Can you send us a support bundle? You can find the link at the footer of UI. After that, can you try to scale down all the Kubernetes workloads to zero? This should result in all the volumes detached. Then there shouldn't be anyone using the engine binary, then the installation should continue.

Btw, in 0.8.0, the directory for the engine binaries should be at `/var/lib/longhorn` instead of `/var/lib/rancher/longhorn` since we've changed the root directory to remove `rancher` part in v0.8.0.

**Comment by dstuart** [comment]
So I ended up fixing it, by going into the correct folder (/var/lib/longhorn) and removing the engine folder for v0.8.0 and redeploying the container. Not sure why or how the engine got corrupted but they look exactly the same with same permissions and same sizes. Would you want a backup of them?

**Comment by yasker** [comment]
@dstuart That's great. Regarding reproducing the issue and fix the root cause, if you can send us a support bundle (even after it's worked around), that would be a great help. You can send it to `longhorn-support-bundle@rancher.com`.

I will keep this issue until we figure out the root cause.

**Comment by shuo-wu** [comment]
Reproduce step:
1. Deploy Longhorn v0.8.0
2. Create and attach a volume
3. Open the volume detail page for the volume in Longhorn UI
4. Delete the related DaemonSet of engine image v0.8.0:
```
~ kubectl -n longhorn-system delete ds engine-image-ei-e10d6bf5
```
5. Check the log of the newly created engine-image pods. They should crash with the error log:
```
~ kubectl -n longhorn-system logs engine-image-ei-e10d6bf5-xxxxx
cp: cannot create regular file '/data/longhorn': Text file busy
```

This bug happens only in v0.8.0. I will dig into it later.

**Comment by yasker** [comment]
`cp -n` will not be enough if the image is e.g. `master` tag.

If there is an existing binary, we can calculate the md5/sha1 and only replace it if it's different from what we're about to install, so at least it won't be an issue during the production deployment (since we will not reuse the tag).

Though I don't understand the first solution.

**Comment by shuo-wu** [comment]
The key point of the 1st solution is decoupling the binary cleanup with the engine image pod/DaemonSet deletion/restart. Since the pod deletion/restart doesn't mean that we need to remove the related binary. The only situation in which we need to do the cleanup is engine image object deletion and the binary is no longer used.

**Comment by yasker** [comment]
But how is it related to this issue? Is the issue caused by `cp` failed or cleanup failed? I assume it's the `cp`?

**Comment by shuo-wu** [comment]
Here is the logic:
The cause of this issue is the restarted/re-launched engine image pod is trying to overwrite the binary --> The binary is not cleaned up correctly during the previous pod shutdown --> The reason of the cleanup failure is due to the binary being used...

OK. It seems that we still need to avoid overwriting the binary with `cp`.

**Comment by yasker** [comment]
Just thought about it, we don't need md5 or sha1. Just run a binary `diff` will be enough.

**Comment by shuo-wu** [comment]
Test case1:
1. Clean up all engine binaries on all nodes before Launching Longhorn:
```
rm -rf /var/lib/longhorn/engine-binaries/*
```
2. Deploy Longhorn.
3. Create a pod with Longhorn volume. Then keep writing data into the volume.
4. Delete the engine image DaemonSet. e.g., For `longhornio/longhorn-engine:master`, it's:
```
kubectl -n longhorn-system delete ds engine-image-ei-605a0f3e
```
5. Check if the data writing still works
6. Wait for the engine image becoming `ready` then check the data and if volume snapshot/backup works fine.

Test case2:
1. Create an invalid/fake engine binary file for each node. If the engine image you are using is `longhornio/longhorn-engine:master`, the creation is like:
```
# touch /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-master/longhorn
# ll /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-master/longhorn
-rw-r--r-- 1 root root 0 Mar 30 05:16 /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-master/longhorn
```
2. Deploy Longhorn.
3. Check if Longhorn can be launched successfully and if Longhorn volume works fine.
4. Check if the file on each node is replaced by a real/valid engine binary
```
# ll /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-master/longhorn
-rwxr-xr-x 1 root root 26304136 Mar 26 06:38 /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-master/longhorn*
# /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-master/longhorn
NAME:
   longhorn - A new cli application

USAGE:
   longhorn [global options] command [command options] [arguments...]

VERSION:
   fdea55d

COMMANDS:
   controller
   replica
   sync-agent
   sync-agent-server-reset
   start-with-replicas, start
   add-replica, add
   ls-replica, ls
   rm-replica, rm
   replica-rebuild-status, rebuild-status
   snapshots, snapshot
   backups, backup
   expand
   journal
   info
   frontend
   version
   help, h                                 Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --url value    (default: "http://localhost:9501")
   --debug
   --help, -h     show help
   --version, -v  print the version
```

**Comment by cyboman32** [comment]
Hey guys. While we are waiting for update with correct systematic fix from longhorn developers, I've managed to make a quick fix. Just patched the `engine-image-ei-xxxxxxx` DaemonSet:

was:
```
...
    spec:
      containers:
      - args:
        - -c
        - cp /usr/local/bin/longhorn* /data/ && echo installed && trap 'rm
          /data/longhorn* && echo cleaned up' EXIT && sleep infinity
        command:
        - /bin/bash
        image: longhornio/longhorn-engine:v0.8.0
...
```

now:
```
...
    spec:
      containers:
      - args:
        - -c
        - cp /usr/local/bin/longhorn* /data/ | true && echo installed && trap 'rm
          /data/longhorn* && echo cleaned up' EXIT && sleep infinity
        command:
        - /bin/bash
        image: longhornio/longhorn-engine:v0.8.0
...
```

So I've skip `cp` error code checking by adding '| true' after cp and longhorn up and run.

**Comment by meldafrawi** [comment]
Validation: **PASSED**

**Comment by miraculixx** [comment]
Run into the same problem, was able to temporarily fix and recover the failed deployment by manually editing the engine-image-xyz daemonset, setting the following command:

```
-c "cp /usr/local/bin/longhorn* /data/ | true && echo installed && trap 'rm /data/longhorn* && echo cleaned up' EXIT && sleep infinity"
```

The part added is the `| true `

**Comment by yasker** [comment]
@miraculixx which version you're using? We should have fixed it since v0.8.1 release.

**Comment by miraculixx** [comment]
@yasker thank you, forgot to mention. Unfortunately release upgrade was not done in time on this cluster.
